### ISSUE TITLE
Capture with `as` in `match` statement breaks type narrowing

### ISSUE THREAD

**Issue by mthuurne** [astral-sh/ty#928]
### Summary

[testcase in playground](https://play.ty.dev/42bb391c-e742-4b01-adf1-7a91cee18a26)

If I check the following code:
```py
from typing import Never, NoReturn


def bad_type(value: Never) -> NoReturn:
    raise TypeError(type(value))


def f(value: str | int) -> None:
    match value:
        case int() as x:
            print("integer:", x)
        case str():
            print("string")
        case _:
            bad_type(value)
```
ty reports:
```
error[invalid-argument-type]: Argument to function `bad_type` is incorrect
  --> narrowing_to_never.py:15:22
   |
13 |             print("string")
14 |         case _:
15 |             bad_type(value)
   |                      ^^^^^ Expected `Never`, found `int`
   |
info: Function defined here
 --> narrowing_to_never.py:4:5
  |
4 | def bad_type(value: Never) -> NoReturn:
  |     ^^^^^^^^ ------------ Parameter declared here
5 |     raise TypeError(type(value))
  |
info: rule `invalid-argument-type` is enabled by default
```
I would expect it to not report any issues, as the default case is unreachable if a properly typed value is passed.

When the capture (`as x`) is removed, no issues are reported. It seems that the capture disables the type narrowing.

### Version

ty 0.0.1-alpha.16

**Comment by mthuurne** [comment]
[testcase in playground](https://play.ty.dev/42bb391c-e742-4b01-adf1-7a91cee18a26)
